### DISCUSSION TITLE
Very Slow Query Search Time and QPS

### DISCUSSION THREAD

**Discussion by majedtaki** [milvus-io/milvus#17500] [Q&A and General discussion]
I deployed a milvus cluster on GKE with SSD for minio,etcd and pulsar. The cluster has 7 nodes where each node has 32CPU/128GB Ram for a total of 128 CPU and 512GB of ram.

I have the cluster scaled up as follows:

4 index pods 
8 query pods -> resource requests of for each pod is (8 cpu and 32 GB)
10 proxy pods

I inserted 50M random 512 dimension vectors. I tried both IVF_PQ, IVF_FLAT indexes (with different nlist sizes (256, 2024, 40k) but I get very bad query speed.

Here are the steps (all done using python sdk):
- Load 50M random vectors in parallel with batch size of 100k. Takes around 20 mins.
- Do a single query search (nq=1, topk=100, nprobe=32) : Latency around 700ms
- Do 10 concurrent query search: Latency is 3 seconds per request

This is clearly very far off from what the official benchmark shows. What am I doing wrong?

**Comment by xiaofan-luan** [comment]
> I deployed a milvus cluster on GKE with SSD for minio,etcd and pulsar. The cluster has 7 nodes where each node has 32CPU/128GB Ram for a total of 128 CPU and 512GB of ram.
> 
> I have the cluster scaled up as follows:
> 
> 4 index pods 8 query pods -> resource requests of for each pod is (8 cpu and 32 GB) 10 proxy pods
> 
> I inserted 50M random 512 dimension vectors. I tried both IVF_PQ, IVF_FLAT indexes (with different nlist sizes (256, 2024, 40k) but I get very bad query speed.
> 
> Here are the steps (all done using python sdk):
> 
> * Load 50M random vectors in parallel with batch size of 100k. Takes around 20 mins.
> * Do a single query search (nq=1, topk=100, nprobe=32) : Latency around 700ms
> * Do 10 concurrent query search: Latency is 3 seconds per request
> 
> This is clearly very far off from what the official benchmark shows. What am I doing wrong?

1) Upgrade to 2.1 which be ready next week, you will see great performance increase
2) Use HNSW and will be much faster in searching(Build Index will be slower)
3) Deploy the monitoring system help you to check if cluster is balanced

**Comment by xiaofan-luan** [comment]
Let us know your use case and performance goal and let's see how we can help you furter

**Reply by majedtaki** [reply]
Actually I got it to run faster. I managed to get 140 QPS on 50M 512 vectors but I couldn't scale it further than that. I understand version 2.1 should have multi replica index which should help with throughput.

However I encountered alot of errors and bugs. When inserting a lot of data, and a datanode or a querynode restarts, the whole system freezes due to this errors: `"[Failed to create producer]" error="server error: ProducerBlockedQuotaExceededException: Cannot create producer on topic with backlog quota exceeded" producerID=7 producer_name=milvus-pulsar-9-9 topic="persistent://public/default/by-dev-rootcoord-dml_2"`

I'll have to ssh into the broker pod and manually clear the backlog for each subscription for that topic in order to unfreeze the system.

Will that issue be fixed with version 2.1 as well?

Thanks!

**Reply by xiaofan-luan** [reply]
> Actually I got it to run faster. I managed to get 140 QPS on 50M 512 vectors but I couldn't scale it further than that. I understand version 2.1 should have multi replica index which should help with throughput.
> 
> However I encountered alot of errors and bugs. When inserting a lot of data, and a datanode or a querynode restarts, the whole system freezes due to this errors: `"[Failed to create producer]" error="server error: ProducerBlockedQuotaExceededException: Cannot create producer on topic with backlog quota exceeded" producerID=7 producer_name=milvus-pulsar-9-9 topic="persistent://public/default/by-dev-rootcoord-dml_2"`
> 
> I'll have to ssh into the broker pod and manually clear the backlog for each subscription for that topic in order to unfreeze the system.
> 
> Will that issue be fixed with version 2.1 as well?
> 
> Thanks!
 
Yep, most of the pulsar unregister issue has been fixed and more over we removed pulsar dependency in the search path, so we should expect much better performance and stability in 2.1. It will be released next week. 

Feel free to open issue if you hit any other problems

**Reply by majedtaki** [reply]
I have another question. How do I scale the query nodes? Looks like the data does not get evenly distributed across the nodes? When I add new query nodes, they don't load any data and will have 0 cpu utilization when I do concurrent queries.

I looked into the query coord and looks like its getting the mem usage wrong?
```
["loadBalanceSegmentLoop: memory usage rate of all online QueryNode"] ["mem rate"="{\"358\":9.497957975668218e-11,\"362\":9.705880543720017e-11,\"363\":1.1922018927634786e-10,\"365\":9.997025429697718e-11,\"366\":1.3478551608159256e-10,\"370\":1.1076872752369132e-10,\"372\":1.329141241512844e-10,\"385\":1.873168287147565e-12,\"390\":1.83764115035956e-12,\"391\":1.7577050925865486e-12,\"392\":1.7852386235972525e-12,\"393\":1.817657135916307e-12}"]
```

**Reply by xiaofan-luan** [reply]
> I have another question. How do I scale the query nodes? Looks like the data does not get evenly distributed across the nodes? When I add new query nodes, they don't load any data and will have 0 cpu utilization when I do concurrent queries.
> 
> I looked into the query coord and looks like its getting the mem usage wrong?
> 
> ```
> ["loadBalanceSegmentLoop: memory usage rate of all online QueryNode"] ["mem rate"="{\"358\":9.497957975668218e-11,\"362\":9.705880543720017e-11,\"363\":1.1922018927634786e-10,\"365\":9.997025429697718e-11,\"366\":1.3478551608159256e-10,\"370\":1.1076872752369132e-10,\"372\":1.329141241512844e-10,\"385\":1.873168287147565e-12,\"390\":1.83764115035956e-12,\"391\":1.7577050925865486e-12,\"392\":1.7852386235972525e-12,\"393\":1.817657135916307e-12}"]
> ```

looks like your memory rate is really low, did you set memory limit on docker?

**Reply by majedtaki** [reply]
@xiaofan-luan No limit is set for resources. The cluster is deployed on google GKE. The query pods are deployed on 8 core 8GB ram node where each pod uses a whole node by itself. I think there might be a bug on how the querycoord fetches the mem usauge?

The only way I got it to load balance correctly is to insert the 50M data, wait for it to be persisted to disk and then call `collection.load()` which then the query nodes load the segment to memory and they are all balanced. However once its loaded, if I change the number of query nodes, the load is not balanced automatically.

However in production, we will be inserting small batches and loading the collection immediately after so it becomes available for search while we update the index

**Reply by xiaofan-luan** [reply]
1.I guess that might be the problem because we balance based on physical machine memory or container memory limit.
 2. if I change the number of query nodes, we should automatically balanced based on memory.
 3. there are two querynodes(If your shards are two) handle growing segments, which might consume a little bit more memory than other querynodes.
If set memory limit won't help please offer querycoord and querynode log and let's investigate into detailes.

**Reply by majedtaki** [reply]
Hi @xiaofan-luan sorry last question. I'm created a 50M vector collection and created 16 partitions as well. However when I try to load it, the fails and I see that the bookie pods restarts with these errors:

```
io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 16777216 byte(s) of direct memory (used: 8589934592, max: 8589934592) at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:802)
```

Any idea whats going on? This only happens when I create a collection with multiple partitions.

**Reply by majedtaki** [reply]
^ UPDATE:  It was a memory issue with pulsar bookie pods. The error has been resolved by increasing the memory resources
